### PR TITLE
#2778699: Updates connection routes to work with all entities

### PR DESCRIPTION
**Pull request by aprice42**
I think I have connection pages and forms working properly for all entities. 

**TODO:** 
- Add better error/empty connection handling
- Update urls on connection pages
- Add default routes and callbacks for connections
- Create view to show connections for specific entities
- Add connections tab/s to entities

### CODE REVIEW

File: modules/redhen_connection/redhen_connection.module
Comment (unclegcb) [inline modules/redhen_connection/redhen_connection.module:168]: Huh. I wonder if we could do a direct DB-query instead? Worth considering as a future optimization perhaps.

File: modules/redhen_connection/src/ConnectionHtmlRouteProvider.php
Comment (unclegcb) [inline modules/redhen_connection/src/ConnectionHtmlRouteProvider.php]: This seems to be the only place that `$entity_type` or its ID get used. I assume this is the redhen connection entity type? But if that's true, why are we looping through every entity type in the system, why not just use the two entity types that are allowed to be linked by the connection type? Maybe I'm confused about what's happening here.

The same thing exists in the next function, `getAddEntityFormRoutes()`, as does a lot of similar logic. I'm wondering if these two could share a helper function to avoid repetition?

File: modules/redhen_connection/src/Controller/ConnectionAddController.php
Comment (unclegcb) [inline modules/redhen_connection/src/Controller/ConnectionAddController.php]: This seems a little risky: looping through some unclear parameters and accepting anything that uses a "type" key. Maybe there's a cleaner way in these suggestions? Or did you try these already? https://www.drupal.org/forum/support/module-development-and-code-questions/2014-07-25/drupal-8-get-entity-object-given

File: modules/redhen_connection/src/Controller/RedhenConnections.php
Comment (unclegcb) [inline modules/redhen_connection/src/Controller/RedhenConnections.php]: Repeat of the code in the other controller: let's put it somewhere we can share it. Even just a function in the .module file.

File: modules/redhen_connection/src/Controller/RedhenConnections.php
Comment (unclegcb) [inline modules/redhen_connection/src/Controller/RedhenConnections.php:11]: I just want to stop and appreciate that you alphabetized your `use` statements. <3

File: modules/redhen_connection/src/Plugin/Derivative/RedhenConnectionLocalTask.php
Comment (unclegcb) [inline modules/redhen_connection/src/Plugin/Derivative/RedhenConnectionLocalTask.php]: This variable isn't doing anything, right? I assume we can remove the assignment and make this more reasonable. It would also become obsolete if we make connections only use field-able entities, I'm guessing?

File: modules/redhen_connection/src/Plugin/Derivative/RedhenConnectionLocalTask.php
Comment (unclegcb) [inline modules/redhen_connection/src/Plugin/Derivative/RedhenConnectionLocalTask.php]: Is it necessary to have 2 routes here? My guess in that this is a copy from another chunk of code that also offered an "edit" sub-tab. I think you can just use the top-level tab above, and drop the content right there.

File: modules/redhen_connection/src/Routing/RouteSubscriber.php
Comment (unclegcb) [inline modules/redhen_connection/src/Routing/RouteSubscriber.php]: Should this just be a "view" permission?

File: src/ParamConverter/RedhenParamConverter.php
Comment (unclegcb) [inline src/ParamConverter/RedhenParamConverter.php:3]: I am so delighted to see these weird things go away

**Review by unclegcb — CHANGES_REQUESTED** [review]
Exciting stuff! Great work sorting through how to make this happen. This is code I can imagine we'll be really happy to be able to reference for future work, too.

I put comments inline: some cleanup, some permissions stuff, and some questions about what exactly things are doing. I didn't jump into a debugger to look closer, but if you think that would better answer some of my questions than you trying to sort it out, let me know!

File: modules/redhen_connection/src/ConnectionHtmlRouteProvider.php
Comment (aprice42) [reply modules/redhen_connection/src/ConnectionHtmlRouteProvider.php]: The foreach here is iterating over each entity type and looking for anything with a canonical link template, then it adds a route for each entity with the url pattern `$canonical_path/connection/add` mapped to the ConnectionAddController::add method. 

Yes that`$entity_type` will always just be `redhen_connection`.  I think I might have just passed that variable along because it was used in the prior logic. but I realize that the title I am setting isn't really useful. Seems like it could change to `Add $type->getLabel() Connection`. And I think I can remove that $entity_type variable altogether, since we know what it should be, and it will hopefully make this a bit more readable. 

Combined both functions into one to make it DRY-er.

File: modules/redhen_connection/src/Controller/ConnectionAddController.php
Comment (aprice42) [reply modules/redhen_connection/src/Controller/ConnectionAddController.php]: Good point here, definitely brittle, and it was actually broken for node to node connection types. Found a better approach in the post you shared. Thanks!

File: modules/redhen_connection/src/Plugin/Derivative/RedhenConnectionLocalTask.php
Comment (aprice42) [reply modules/redhen_connection/src/Plugin/Derivative/RedhenConnectionLocalTask.php]: Good point. This is getting replaced because I realized that this logic needed a little more attention to ensure that we only show the connections tab on entities that can have connections.

File: modules/redhen_connection/src/Plugin/Derivative/RedhenConnectionLocalTask.php
Comment (aprice42) [reply modules/redhen_connection/src/Plugin/Derivative/RedhenConnectionLocalTask.php]: Yes, good catch! Thanks!

File: modules/redhen_connection/redhen_connection.module
Comment (aprice42) [reply modules/redhen_connection/redhen_connection.module:168]: Yes, would definitely like to discuss this more.

**Comment by aprice42** [comment]
@unclegcb I made the requested updates and responded to comments. Happy to discuss anything if you like.

File: modules/redhen_connection/src/ConnectionHtmlRouteProvider.php
Comment (unclegcb) [inline modules/redhen_connection/src/ConnectionHtmlRouteProvider.php]: Nice drying out here. Now I could see it going even further and having "getAddEntityRoutes" return a nested array with [form] and [add]! but that's not really helping us much ;)

File: modules/redhen_connection/src/ConnectionHtmlRouteProvider.php
Comment (unclegcb) [inline modules/redhen_connection/src/ConnectionHtmlRouteProvider.php]: Thank you for the variable names: this is *much* more grok-able!

File: modules/redhen_connection/src/ConnectionService.php
Comment (unclegcb) [inline modules/redhen_connection/src/ConnectionService.php:288]: It took me a second to figure out what this was for, but it seems great! I can imagine on a system with a lot of connection types, this function would benefit from just querying the DB directly... but that seems like a very unlikely use case.

Could we be using this in getAddEntityRoutes() instead of `\Drupal::entityTypeManager()->getDefinitions();`? Or as a way to filter those down? Would that get rid of the extra tabs on entity types that don't have connections?

File: modules/redhen_connection/src/Controller/RedhenConnections.php
Comment (unclegcb) [inline modules/redhen_connection/src/Controller/RedhenConnections.php:77]: Let's initialize this outside the `foreach()` so we don't get a PHP warning if there are no connections for a user (on line 88)

File: modules/redhen_connection/src/Controller/RedhenConnections.php
Comment (unclegcb) [reply modules/redhen_connection/src/Controller/RedhenConnections.php]: Looks like this can get deleted now?

File: modules/redhen_connection/src/Entity/ConnectionType.php
Comment (unclegcb) [inline modules/redhen_connection/src/Entity/ConnectionType.php]: This throws a warning because `reset()` takes a reference to a variable, which means you aren't supposed to pass a function into it ("Notice: Only variables should be passed by reference..."). Unfortunately, there's not a great way to still fit this on 1 line, but certainly we don't want to trigger warnings!

Comment (unclegcb) [reply modules/redhen_connection/src/Entity/ConnectionType.php]: (since this is boring, I submitted a fix)

File: modules/redhen_connection/src/Plugin/Derivative/RedhenConnectionLocalTask.php
Comment (unclegcb) [inline modules/redhen_connection/src/Plugin/Derivative/RedhenConnectionLocalTask.php]: This is an aggressive default! Really wanna be last, huh? I might drop it to 20.

File: modules/redhen_connection/src/Routing/RouteSubscriber.php
Comment (unclegcb) [inline modules/redhen_connection/src/Routing/RouteSubscriber.php]: I think I'm gonna opt for yanking the branding out of the title here.

File: modules/redhen_connection/src/Controller/RedhenConnections.php
Comment (unclegcb) [reply modules/redhen_connection/src/Controller/RedhenConnections.php:77]: (Done)

File: modules/redhen_connection/src/Plugin/Derivative/RedhenConnectionLocalTask.php
Comment (unclegcb) [reply modules/redhen_connection/src/Plugin/Derivative/RedhenConnectionLocalTask.php]: (Done)

File: modules/redhen_connection/src/Routing/RouteSubscriber.php
Comment (unclegcb) [reply modules/redhen_connection/src/Routing/RouteSubscriber.php]: (Done)

**Review by unclegcb — COMMENTED** [review]
@aprice42 this is looking great! I have a couple little tweaks that I went ahead and committed (all commented inline) and one suggested deletion that I wanted to verify: then I think we're good to merge! Let me know if I'm missing something about that method.

**Comment by unclegcb** [comment]
@aprice42 when you look at this next -- and I hope it won't need much, as I had little feedback -- can you glance at https://www.drupal.org/project/redhen/issues/3025703 and tell me if it's good, bad, or needs some time to look into? Same with https://www.drupal.org/project/redhen/issues/3004137

File: modules/redhen_connection/src/ConnectionHtmlRouteProvider.php
Comment (aprice42) [reply modules/redhen_connection/src/ConnectionHtmlRouteProvider.php]: Good call on this. I went ahead and changed it. :-)

File: modules/redhen_connection/src/ConnectionService.php
Comment (aprice42) [reply modules/redhen_connection/src/ConnectionService.php:288]: done!

File: modules/redhen_connection/src/Controller/RedhenConnections.php
Comment (aprice42) [reply modules/redhen_connection/src/Controller/RedhenConnections.php]: right! good call. 😄

File: modules/redhen_connection/src/Plugin/Derivative/RedhenConnectionLocalTask.php
Comment (aprice42) [reply modules/redhen_connection/src/Plugin/Derivative/RedhenConnectionLocalTask.php]: 😄

**Comment by aprice42** [comment]
@unclegcb requested changes have been made. 

Looking at the two issues you mentioned: 
https://www.drupal.org/project/redhen/issues/3004137 needs more work, the patch just comments out the broken logic.

https://www.drupal.org/project/redhen/issues/3025703 looking through the patch, it looks like it could be good, but I would want to spend 0.5-1h reviewing to make sure it works as expected.

File: modules/redhen_connection/redhen_connection.module
Comment (unclegcb) [inline modules/redhen_connection/redhen_connection.module:175]: WHOOMP there it is!!

**Review by unclegcb — APPROVED** [review]
Great work @aprice42 ! I'm gonna run some more local tests before merging, but I expect that to all go according to plan. Thanks for sticking with this task!

**Pull request merged**
